sympy/matrices/eigen.py

```diff
@@ -33,32 +33,30 @@ def _eigenvals_triangular(M, multiple=False):
 
 
 def _eigenvals_eigenvects_mpmath(M):
-    mp = MPContext()
-    prec = max([x._prec for x in M.atoms(Float)])
-
-    mp.prec = prec
-    A = mp.matrix(M.evalf(n=prec_to_dps(prec)))
-    E, ER = mp.eig(A)
-    norm = mp.fsum([mp.re(e)**2 + mp.im(e)**2 for e in E])
-
-    while True:
-        prec_new = round(prec * 1.5)
-
-        if prec_new > DEFAULT_MAXPREC:
-            raise PrecisionExhausted
+    norm2 = lambda v1, v2: max(
+        [max(mp.fabs(r1 - r2), mp.fabs(i1 - i2))
+         for (r1, i1), (r2, i2) in zip(v1, v2)])
 
-        mp.prec = prec_new
-        A_new = mp.matrix(M.evalf(n=prec_to_dps(prec)))
-        E_new, ER_new = mp.eig(A_new)
-        norm_new = mp.fsum([mp.re(e)**2 + mp.im(e)**2 for e in E_new])
-
-        if abs(norm - norm_new) < max(norm, norm_new) * 2**-prec:
-            break
-
-        norm = norm_new
-        E, ER = E_new, ER_new
-
-    return E, ER
+    v1 = None
+    prec = max([x._prec for x in M.atoms(Float)])
+    eps = 2**-prec
+
+    while prec < DEFAULT_MAXPREC:
+        with workprec(prec):
+            A = mp.matrix(M.evalf(n=prec_to_dps(prec)))
```

Comment:
I think if you leave the `n=` arg off, it will use the working precision as defined in the `with` context`.